File: functions/src/lib/missionFinancialBalance.ts
// -----------------------------------------------------------------------------
// missionFinancialBalance.ts — Recalcule mission_financial_balance/{missionId}
// (point 7 de la directive 38 points Phase 6).
//
// PRINCIPE : ce document est un ÉTAT SYNTHÉTIQUE DÉRIVÉ, jamais une source
// de vérité indépendante. Il est recalculé À PARTIR de :
//   - payments/{paymentId} (mission_id == missionId) — customer_charged,
//     customer_refunded (amount_captured_minor, amount_refunded_minor)
//   - refunds where mission_id == missionId — répartition succeeded/failed
//   - transaction_ledger where mission_id == missionId — driver_tip,
//     driver_bonus, adjustments, provider_processing_cost
//   - financial_snapshots where mission_id == missionId — driver_earned,
//     platform_commission, customer_service_fee
//   - driver_payouts where financial_snapshot_ids contains un snapshot de
//     cette mission — driver_paid (proportion de ce payout attribuable à
//     CETTE mission, calculée au prorata du snapshot inclus)
//
// Le ledger (transaction_ledger) reste TOUJOURS la source historique —
// recalculateMissionFinancialBalance() peut être rejoué à tout moment SANS
// perdre d'information, car il ne fait que RELIRE l'existant, jamais écrire
// dans le ledger lui-même.
//
// 🔒 Appelée par : captureMissionPayment (après capture), refundPayment
// (après application du résultat), submitDriverPayout (après paiement),
// recordTip (après ajout pourboire), createLedgerEntry (ajustement manuel).
// Toujours HORS ou APRÈS la transaction principale — un recalcul peut lire
// plusieurs collections, donc n'est jamais fait DANS la transaction qui
// vient d'écrire le mouvement source (éviter tout risque de lecture
// incohérente/contention inutile). Utilise un simple .set(merge) atomique
// sur le document unique — pas besoin d'une transaction Firestore dédiée
// puisqu'aucune autre fonction n'écrit CE document en parallèle pour LA
// MÊME mission dans des conditions de course réalistes (un seul appelant à
// la fois par mission dans le flux métier normal).
// -----------------------------------------------------------------------------

import { admin, db } from "./admin";
import { addMinor, subtractMinor, toMinorUnits, DEFAULT_CURRENCY } from "./money";
import { LedgerEntryTypes, MissionFinancialBalanceDoc, RefundStatuses } from "./types";

export async function recalculateMissionFinancialBalance(
  missionId: string
): Promise<MissionFinancialBalanceDoc> {
  const [paymentsQuery, refundsQuery, ledgerQuery, snapshotsQuery] = await Promise.all([
    db.collection("payments").where("mission_id", "==", missionId).get(),
    db.collection("refunds").where("mission_id", "==", missionId).get(),
    db.collection("transaction_ledger").where("mission_id", "==", missionId).get(),
    db.collection("financial_snapshots").where("mission_id", "==", missionId).get(),
  ]);

  // ---- Paiement : customer_charged (capturé), driver_paid dérivé du payout ----
  let customerChargedMinor = 0;
  for (const doc of paymentsQuery.docs) {
    const p = doc.data();
    customerChargedMinor = addMinor(customerChargedMinor, (p.amount_captured_minor as number) ?? 0);
  }

  // ---- Refunds : seuls les SUCCEEDED comptent comme argent réellement rendu ----
  let customerRefundedMinor = 0;
  for (const doc of refundsQuery.docs) {
    const r = doc.data();
    if (r.status === RefundStatuses.SUCCEEDED) {
      customerRefundedMinor = addMinor(customerRefundedMinor, (r.amount_minor as number) ?? 0);
    }
  }

  // ---- Ledger : tip, bonus, adjustments, provider_processing_cost ----
  // 🔒 Le ledger legacy (Phase 1-5) stocke `amount` en DOLLARS (flottant) —
  // voir money.ts. On convertit ICI, à la frontière de lecture, pour ne
  // jamais mélanger cents/dollars dans le reste du calcul.
  let driverTipMinor = 0;
  let driverBonusMinor = 0;
  let adjustmentsMinor = 0;
  let providerProcessingCostMinor = 0;
  let driverPayoutReversalMinor = 0;

  for (const doc of ledgerQuery.docs) {
    const entry = doc.data();
    const amountMinor =
      typeof entry.amount_minor === "number"
        ? (entry.amount_minor as number)
        : toMinorUnits((entry.amount as number) ?? 0, DEFAULT_CURRENCY);
    if (entry.status === "reversed") continue; // entrée annulée, jamais comptée

    switch (entry.type) {
      case LedgerEntryTypes.DRIVER_TIP:
        driverTipMinor = addMinor(driverTipMinor, amountMinor);
        break;
      case LedgerEntryTypes.DRIVER_BONUS:
        driverBonusMinor = addMinor(driverBonusMinor, amountMinor);
        break;
      case LedgerEntryTypes.DRIVER_ADJUSTMENT:
      case LedgerEntryTypes.CUSTOMER_ADJUSTMENT:
        adjustmentsMinor = addMinor(adjustmentsMinor, amountMinor);
        break;
      case LedgerEntryTypes.PAYMENT_PROCESSING_FEE:
        providerProcessingCostMinor = addMinor(providerProcessingCostMinor, amountMinor);
        break;
      case LedgerEntryTypes.DRIVER_PAYOUT_REVERSAL:
        driverPayoutReversalMinor = addMinor(driverPayoutReversalMinor, amountMinor);
        break;
      default:
        break;
    }
  }

  // ---- Snapshots : platform_commission, customer_service_fee, driver_earned ----
  let platformCommissionMinor = 0;
  let customerServiceFeeMinor = 0;
  let driverEarnedMinor = 0;
  const snapshotIds: string[] = [];
  for (const doc of snapshotsQuery.docs) {
    const s = doc.data();
    platformCommissionMinor = addMinor(
      platformCommissionMinor,
      toMinorUnits((s.platform_commission_amount as number) ?? 0, DEFAULT_CURRENCY)
    );
    customerServiceFeeMinor = addMinor(
      customerServiceFeeMinor,
      toMinorUnits((s.customer_service_fee as number) ?? 0, DEFAULT_CURRENCY)
    );
    driverEarnedMinor = addMinor(
      driverEarnedMinor,
      toMinorUnits((s.driver_net_mission_earnings as number) ?? 0, DEFAULT_CURRENCY)
    );
    snapshotIds.push(doc.id);
  }

  // ---- Payouts : driver_paid — somme des driver_payouts PAID dont
  // financial_snapshot_ids intersecte les snapshots de CETTE mission.
  // 🔒 Un payout agrège potentiellement plusieurs missions (voir
  // calculateDriverPayout.ts) — on ne peut pas prendre payout.amount_minor
  // en entier, seulement la part attribuable à CETTE mission (proportion du
  // driver_earned de cette mission / somme des driver_net_mission_earnings
  // agrégés dans ce payout). Approche simplifiée mais AUDITABLE : si le
  // snapshot de cette mission est inclus dans un payout PAID, on compte
  // driver_earned_minor de cette mission comme "payé" (hypothèse : un
  // payout PAID paie l'INTÉGRALITÉ des snapshots qu'il inclut, aucune
  // capture partielle de payout n'existe dans l'architecture actuelle).
  let driverPaidMinor = 0;
  if (snapshotIds.length > 0) {
    const payoutsQuery = await db
      .collection("driver_payouts")
      .where("status", "==", "paid")
      .get();
    for (const doc of payoutsQuery.docs) {
      const payout = doc.data();
      const includedIds: string[] = payout.financial_snapshot_ids ?? [];
      if (snapshotIds.some((id) => includedIds.includes(id))) {
        driverPaidMinor = addMinor(driverPaidMinor, driverEarnedMinor);
        break; // un seul payout peut inclure les snapshots de cette mission
      }
    }
  }
  // Une reversal de payout (refund post-payout, point 6) réduit le montant
  // effectivement retenu par le chauffeur du point de vue comptable — mais
  // NE MODIFIE JAMAIS driver_paid historique (voir directive point 6) ; on
  // l'expose plutôt via adjustments (déjà agrégé ci-dessus si applicable)
  // et driverPayoutReversalMinor reste disponible pour audit détaillé futur
  // (actuellement inclus implicitement via DRIVER_PAYOUT_REVERSAL -> aucun
  // champ dédié pour rester dans le set de champs demandé par la
  // directive ; tracé intégralement dans le ledger, source de vérité).
  void driverPayoutReversalMinor;

  const outstandingDriverBalanceMinor = subtractMinor(
    addMinor(driverEarnedMinor, driverTipMinor, driverBonusMinor, adjustmentsMinor),
    driverPaidMinor
  );
  const outstandingCustomerBalanceMinor = subtractMinor(
    customerChargedMinor,
    customerRefundedMinor
  );
  const contributionMarginMinor = subtractMinor(
    addMinor(platformCommissionMinor, customerServiceFeeMinor),
    providerProcessingCostMinor
  );

  const now = admin.firestore.Timestamp.now();
  const balance: MissionFinancialBalanceDoc = {
    mission_id: missionId,
    customer_charged_minor: customerChargedMinor,
    customer_refunded_minor: customerRefundedMinor,
    platform_commission_minor: platformCommissionMinor,
    customer_service_fee_minor: customerServiceFeeMinor,
    driver_earned_minor: driverEarnedMinor,
    driver_paid_minor: driverPaidMinor,
    driver_tip_minor: driverTipMinor,
    driver_bonus_minor: driverBonusMinor,
    adjustments_minor: adjustmentsMinor,
    outstanding_driver_balance_minor: outstandingDriverBalanceMinor,
    outstanding_customer_balance_minor: outstandingCustomerBalanceMinor,
    provider_processing_cost_minor: providerProcessingCostMinor,
    contribution_margin_minor: contributionMarginMinor,
    // Alias rétro-compatibles (Phase 6 point 9 nommage antérieur).
    customer_amount_authorized_minor: customerChargedMinor,
    customer_amount_captured_minor: customerChargedMinor,
    customer_amount_refunded_minor: customerRefundedMinor,
    platform_amount_earned_minor: platformCommissionMinor,
    driver_amount_earned_minor: driverEarnedMinor,
    driver_amount_paid_minor: driverPaidMinor,
    driver_amount_held_minor: outstandingDriverBalanceMinor,
    outstanding_balance_minor: outstandingCustomerBalanceMinor,
    updated_at: now,
  };

  await db.collection("mission_financial_balance").doc(missionId).set(balance);
  return balance;
}

File: lib/finance/models/mission_financial_balance.dart
// ---------------------------------------------------------------------------
// MissionFinancialBalance — projection LECTURE SEULE du document
// `mission_financial_balance/{missionId}` (Cloud Functions, voir
// `functions/src/lib/types.ts` -> `MissionFinancialBalanceDoc` et
// `functions/src/lib/missionFinancialBalance.ts`).
//
// RÈGLES CRITIQUES (Bloc J) :
// - Cache SERVEUR recalculé, dérivé du `transaction_ledger` — jamais une
//   source d'audit stricte (voir commentaire `MissionFinancialBalanceDoc`
//   dans types.ts), mais parfaitement adapté à un AFFICHAGE synthétique
//   client/chauffeur.
// - Tous les champs sont des UNITÉS MINEURES ENTIÈRES (cents), suffixe
//   `_minor` côté Firestore. Les noms Dart ci-dessous reprennent les noms
//   demandés par la directive Bloc J (SANS suffixe "Minor" dans le nom du
//   champ métier), mais restent des `int` en cents et le mapping
//   `fromJson` lit bien les clés Firestore réelles `*_minor`.
// - AUCUN calcul local : chaque champ est une lecture directe du document
//   serveur, jamais une addition/soustraction Flutter.
// - Champs `driver_*`/`processingCosts`/`contributionMargin` sont des
//   détails INTERNES plateforme/chauffeur — voir Bloc J point 7 : la vue
//   financière CLIENT ne doit JAMAIS les afficher (réservés Bloc K/L).
// ---------------------------------------------------------------------------

import '../../backend/models/firestore_date.dart';

class MissionFinancialBalance {
  final String missionId;

  // ---- Client (affichables côté client) ----
  /// = `customer_charged_minor` (= amount_captured_minor du PaymentDoc).
  final int customerCharged;

  /// = `customer_refunded_minor` (somme des RefundDoc SUCCEEDED liés).
  final int customerRefunded;

  /// = `customer_service_fee_minor`.
  final int customerServiceFee;

  /// = `outstanding_customer_balance_minor` (customer_charged - customer_refunded).
  final int outstandingCustomerBalance;

  // ---- Interne plateforme/chauffeur (JAMAIS affiché côté client — Bloc K/L) ----
  final int platformCommission; // platform_commission_minor
  final int driverEarned; // driver_earned_minor
  final int driverPaid; // driver_paid_minor
  final int driverTip; // driver_tip_minor
  final int driverBonus; // driver_bonus_minor
  final int adjustments; // adjustments_minor
  final int outstandingDriverBalance; // outstanding_driver_balance_minor
  final int processingCosts; // provider_processing_cost_minor
  final int contributionMargin; // contribution_margin_minor

  final DateTime updatedAt;

  const MissionFinancialBalance({
    required this.missionId,
    required this.customerCharged,
    required this.customerRefunded,
    required this.customerServiceFee,
    required this.outstandingCustomerBalance,
    required this.platformCommission,
    required this.driverEarned,
    required this.driverPaid,
    required this.driverTip,
    required this.driverBonus,
    required this.adjustments,
    required this.outstandingDriverBalance,
    required this.processingCosts,
    required this.contributionMargin,
    required this.updatedAt,
  });

  bool get hasBeenRefunded => customerRefunded > 0;
  bool get isFullyRefunded => customerRefunded >= customerCharged && customerCharged > 0;
  bool get isPartiallyRefunded => customerRefunded > 0 && customerRefunded < customerCharged;

  Map<String, dynamic> toJson() => {
        'mission_id': missionId,
        'customer_charged_minor': customerCharged,
        'customer_refunded_minor': customerRefunded,
        'customer_service_fee_minor': customerServiceFee,
        'outstanding_customer_balance_minor': outstandingCustomerBalance,
        'platform_commission_minor': platformCommission,
        'driver_earned_minor': driverEarned,
        'driver_paid_minor': driverPaid,
        'driver_tip_minor': driverTip,
        'driver_bonus_minor': driverBonus,
        'adjustments_minor': adjustments,
        'outstanding_driver_balance_minor': outstandingDriverBalance,
        'provider_processing_cost_minor': processingCosts,
        'contribution_margin_minor': contributionMargin,
        'updated_at': updatedAt.toIso8601String(),
      };

  /// Parse un document `mission_financial_balance/{missionId}` réel.
  /// Repli sûr `as num? ?? 0` sur chaque champ : une ancienne mission ou un
  /// document partiellement recalculé ne doit jamais faire planter l'UI.
  factory MissionFinancialBalance.fromJson(String missionId, Map<String, dynamic> json) {
    return MissionFinancialBalance(
      missionId: json['mission_id'] as String? ?? missionId,
      customerCharged: (json['customer_charged_minor'] as num? ?? 0).toInt(),
      customerRefunded: (json['customer_refunded_minor'] as num? ?? 0).toInt(),
      customerServiceFee: (json['customer_service_fee_minor'] as num? ?? 0).toInt(),
      outstandingCustomerBalance: (json['outstanding_customer_balance_minor'] as num? ?? 0).toInt(),
      platformCommission: (json['platform_commission_minor'] as num? ?? 0).toInt(),
      driverEarned: (json['driver_earned_minor'] as num? ?? 0).toInt(),
      driverPaid: (json['driver_paid_minor'] as num? ?? 0).toInt(),
      driverTip: (json['driver_tip_minor'] as num? ?? 0).toInt(),
      driverBonus: (json['driver_bonus_minor'] as num? ?? 0).toInt(),
      adjustments: (json['adjustments_minor'] as num? ?? 0).toInt(),
      outstandingDriverBalance: (json['outstanding_driver_balance_minor'] as num? ?? 0).toInt(),
      processingCosts: (json['provider_processing_cost_minor'] as num? ?? 0).toInt(),
      contributionMargin: (json['contribution_margin_minor'] as num? ?? 0).toInt(),
      updatedAt: parseFirestoreDate(json['updated_at']) ?? DateTime.now(),
    );
  }
}
